### PR TITLE
[pytorch] split out trace type generator and migrate to new codegen model

### PR DESCRIPTION
**Pull request by ljk53**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#47438 [pytorch] split out trace type generator and migrate to new codegen model**
* #47437 [pytorch] factor out return_names api
* #47436 [pytorch] TraceType codegen tweak - newline before redispatch call

Differential Revision: [D24808211](https://our.internmc.facebook.com/intern/diff/D24808211)

### CODE REVIEW

**Review by bhosmer — APPROVED** [review]
This is a nice chunk of work - will be much easier to rationalize the codegen itself now that it's based on the new data model and moved to its own module. The inline comments are all minor - given byte for byte compatibility, I don't think there's much reason to do heavy rationalization of the codegen logic itself yet (beyond the direct benefits of the data model, that is), since most nontrivial moves would definitely touch the output.

File: tools/autograd/gen_trace_type.py [hunk 1]
```diff
@@ -0,0 +1,412 @@
...
+    # Operations involving Storage or Type are not traceable at the moment
+    if any(str(arg.type) in {'Storage', 'Type', 'ConstQuantizerPtr'}
+           for arg in f.func.schema_order_arguments()):
+        return False
+    # We can't trace functions which don't have any Tensor or TensorList returns
+    if not any(r.type.is_tensor_like() for r in f.func.returns):
+        return False
+    name = cpp.name(f.func)
+    base_name = f.func.name.name.base
+    if base_name in DONT_RECORD_TRACE or name in DONT_RECORD_TRACE:
```
[the remark this answers is not in the document]
Comment (ljk53) [reply tools/autograd/gen_trace_type.py:34]: changed to check `base_name` (also changed `_coalesced_` to `_coalesced`)

File: tools/autograd/gen_trace_type.py [hunk 2]
```diff
@@ -0,0 +1,412 @@
...
+        trace_name = str(f.func.name.name)
+        trace_name = RENAME_TRACE.get(trace_name, trace_name)
+        return OP_NAME.substitute(trace_name=trace_name)
+
+    # otherwise, this is an in-place op and we need to emit both in- and
+    # out-of-place versions
+    outplace_trace_name = f.func.name.name.base
+    inplace_trace_name = cpp.name(f.func)
+    outplace_trace_name = RENAME_TRACE.get(outplace_trace_name, outplace_trace_name)
+    inplace_trace_name = RENAME_TRACE.get(inplace_trace_name, inplace_trace_name)
```
[the remark this answers is not in the document]
Comment (ljk53) [reply tools/autograd/gen_trace_type.py:71]: You mean both `outplace_trace_name` and `inplace_trace_name` or just `inplace_trace_name`? Only `inplace_trace_name` never matches `RENAME_TRACE` - `outplace_trace_name` / `trace_name` above still can.

TBH I'm not 100% sure why the old code names it `outplace_trace_name` - it's used on the `if (force_outplace)` branch below but it's clearly derived from the base name (the old code called utils.uninplace_api_name which removes `_out` suffix manually).

File: tools/codegen/gen.py [hunk 3]
```diff
@@ -821,9 +821,10 @@ def write_with_template(self, filename: str, template_fn: str,
             env = env_callable()
             if isinstance(env, dict):
                 # TODO: Update the comment reference to the correct location
-                comment = "@" + "generated by aten/src/ATen/gen.py"
-                comment += " from {}".format(os.path.basename(template_fn))
-                env['generated_comment'] = comment
+                if 'generated_comment' not in env:
```
[the remark this answers is not in the document]
Comment (ljk53) [reply tools/codegen/gen.py:824]: donno - @ezyang what does 'correct location' mean? full path?

File: tools/autograd/gen_autograd.py [hunk 4]
```diff
@@ -243,6 +243,10 @@ def is_operator_selected_for_training(decl):
         from .gen_variable_type import gen_variable_type
         gen_variable_type(out, aten_decls, template_path)
 
+        from . import gen_trace_type
```
[the remark this answers is not in the document]
Comment (ljk53) [reply tools/autograd/gen_autograd.py:246]: Where can I learn more about the `==` perf regression? Does the lazy import only regress the first run (i.e. benchmark measure problem) or does it actually regress every run (poor implementation of lazy import of python)?
Lazy import might have the benefit of not paying all the cost at cold start time - I don't know if it's desired in some cases for the python frontend code. (hmm, I agree even if so people probably didn't do so deliberately to hide code start latency - they might simply want to localize the imports :-) )

Any way, the codegen script only runs at build time and this is not on hot path so I don't think perf matters here. Unless we want to enforce the top-level import throughout the python codebase for the eager mode frontend code where perf matters.

**Comment by ezyang** [comment]
just added @bdhirsh for visibility, since Brian's been looking at this code recently

**Pull request closed without merging**